forms-flow-components/src/components/CustomComponents/HistoryModal.tsx

```diff
@@ -0,0 +1,346 @@
+    const loadMoreRef = useRef<HTMLDivElement>(null);
+    const lastEntryRef = useRef<HTMLDivElement>(null);
+
+    const handleRevertClick = (
+      version: string,
+      cloned_form_id: string,
+      process_id: string
+    ) => {
+      setSelectedVersion(version);
+      setClonedFormId(cloned_form_id);
+      setProcessId(process_id);
+      setShowConfirmModal(true);
+      onClose();
+    };
+
+    const handleClose = () => {
+      onClose();
+      setHasLoadedMoreForm(false);
+      setHasLoadedMoreWorkflow(false);
+    };
+
+    const handleKeepLayout = () => {
+      setShowConfirmModal(false);
+    };
+
+    const handleReplaceLayout = () => {
+      const idToRevert = categoryType === "FORM" ? clonedFormId : processId;
+      revertBtnAction(idToRevert);
+      setShowConfirmModal(false);
+      setHasLoadedMoreForm(false);
+      setHasLoadedMoreWorkflow(false);
+    };
+
+    const adjustTimelineHeight = () => {
+      const timelineElement = timelineRef.current;
+      const loadMoreElement = loadMoreRef.current;
+      const lastEntryElement = lastEntryRef.current;
+
+      if (timelineElement && lastEntryElement) {
+        const lastEntryHeight = lastEntryElement.offsetHeight;
+        if (loadMoreElement) {
+          const loadMoreHeight = loadMoreElement.offsetHeight;
+          const newHeight = `calc(100% - ${
+            loadMoreHeight + lastEntryHeight
+          }px)`;
+          timelineElement.style.height = newHeight;
+        } else {
+          const newHeight = `calc(100% - ${lastEntryHeight}px)`;
+          timelineElement.style.height = newHeight;
+        }
+      }
+    };
+
+    useEffect(() => {
+      adjustTimelineHeight();
+      window.addEventListener("resize", adjustTimelineHeight);
+      if (show) {
+        if (!hasLoadedMoreForm && categoryType === "FORM") {
+          setHasLoadedMoreForm(false);
+        } else if (!hasLoadedMoreWorkflow && categoryType === "WORKFLOW") {
+          setHasLoadedMoreWorkflow(false);
+        }
+      }
+
+      return () => {
+        window.removeEventListener("resize", adjustTimelineHeight);
+      };
+    }, [show, allHistory]);
+
+    const handleLoadMore = () => {
+      loadMoreBtnAction();
+      if (categoryType === "FORM") {
+        setHasLoadedMoreForm(true);
+      } else if (categoryType === "WORKFLOW") {
+        setHasLoadedMoreWorkflow(true);
+      }
+    };
+
+    const renderHistory = () => {
+      return allHistory.map((entry, index) => {
+        const isMajorVersion = entry.isMajor;
+        const version = `${entry.majorVersion}.${entry.minorVersion}`;
+        const cloned_form_id =
+          categoryType === "FORM" ? entry.changeLog.cloned_form_id : null;
+        const process_id = categoryType === "WORKFLOW" ? entry.id : null;
+        const isLastEntry = index === allHistory.length - 1;
+        return (
+          <React.Fragment key={`${entry.version}-${index}`}>
+            {isMajorVersion && (
+              <div
```

Comment:
From line 200 to 263, can we reduce the duplication . 
Use any AI tools if needed to handle this kinds of optimisations. but make sure you review it and add @Ajay-aot

Reply:
updated